keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1253 @@
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+Running a large computational workload, comprising thousands of pods on
+thousands of nodes requires usage of pod restart policies in order
+to account for infrastructure failures.
+
+Currently, kubernetes Job API offers a way to account for infrastructure
+failures by setting `.backoffLimit > 0`. However, this mechanism intructs the
+job controller to restart all failed pods - regardless of the root cause
+of the failures. Thus, in some scenarios this leads to unnecessary
+restarts of many pods, resulting in a waste of time and computational
+resources. What makes the restarts more expensive is the fact that the
+failures may be encountered late in the execution time of a program.
+
+Sometimes it can be determined from containers exit codes
+that the root cause of a failure is in the executable and the
+job is destined to fail regardless of the number of retries. However, since
+the large workloads are often scheduled to run over night or over the
+weekend, there is no human assistance to terminate such a job early.
+
+The need for solving the problem has been emphasized by the kubernetes
+community in the issues, see: [#17244](https://github.com/kubernetes/kubernetes/issues/17244) and [#31147](https://github.com/kubernetes/kubernetes/issues/31147).
+
+Some third-party frameworks have implemented retry policies for their pods:
+- [TensorFlow Training (TFJob)](https://www.kubeflow.org/docs/components/training/tftraining/)
+- [Argo workflow](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml) ([example](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml))
+
+Additionally, some pod failures are not linked with the container execution,
+but rather with the internal kubernetes cluster management (see:
+[Scheduling, Preemption and Eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/)).
+Such pod failures should be recongnized as infrastructure failures and it
+should be possible to ignore them from the counter towards `backoffLimit`.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Extension of Job API with user-friendly syntax to terminate jobs based on the
+  end state of the failed pod/container.
```

Comment:
Even before the other proposal (#3339), I wondered do we want to go all the way down to containers. Why not just focus on pod level failures, for starters?

Reply:
The idea of using container exit codes is to satisfy the requirement raised in issues: https://github.com/kubernetes/kubernetes/issues/17244 or https://github.com/kubernetes/kubernetes/issues/31147.

Some frameworks like TF or Argo support retries based on container exit codes so it seems reasonable to support it natively. 

Also, having two retry strategies: based on exit codes and pod termination reasons can hopefully lead to a more flexible JobSpec API.